### ISSUE TITLE
Option to download plain, uncompressed files

### ISSUE THREAD

**Issue by w3bb** [int-72h/ofatomic#10]
While compression is good for people with slow connections but good CPUs, it can have detrimental effects for people with slower CPUs but good connections, or people with good CPUs who have the bandwidth to download faster than their CPU can handle . The below graph is from [murse](https://git.sr.ht/~welt/murse), a sort-of reimplementation of ofatomic in Go for my server. Keep in mind the pure-Go implmentation of zstd I use is about 80%-90% performance of the C counterpart. Although the numbers won't be identical, it should give us a rough idea of the CPU usage. 

Below is a part of the CPU profiling graph. This is the two hot paths for downloading files.

![](https://files.catbox.moe/58jfou.png)

As you can see, significantly more time is taken up by zstd than even hashing the files on average. Disabling zstd could have the same performance impact if not more than outright disabling hashing.

**Comment by int-72h** [comment]
Amazing work!
This was an issue I thought would be negligable but it appears I was mistaken.
The issue is some on the team want features such as these to freeze as stainless is being developed but since this is a relatively minor change on client side and on the launcher packer I'll see if I can get it in.
BTW, we may be switching to CSV instead of an sqlite database since it seems to speed up packing times and reduces lib dependencies on stainless.

**Comment by int-72h** [comment]
oh, and the "backend" is just the svn packed with a tool, namely https://github.com/openfortress/launcher-packer

**Comment by w3bb** [comment]
@int-72h Thanks I'll take a look

**Comment by int-72h** [comment]
> @int-72h You should look into using JSON. A lot of the encoders/decoders are extremely well optimised, and many libraries support it. Also it's an absolutely tiny format with little overhead.
> 
> The way murse is set up is storage agnostic (we don't store a database locally, we keep the installed revision number and use to to get the difference between two given revisions). It would just be a matter of hooking it up to the CSV file.
> 
> I definitely think that ofatomic shouldn't be frozen. There are a few problems. I don't want to issue-spam you, but here's some things to think about:
> 
> * You only need to check the signature of the database, and then compare the hash to the database instead of verifying the signature of every file. In effect disabling signature checking would only disable  signature checking for the database.
> * The files aren't truly atomic as you can't download older revisions. There is a carpet-pulling effect that happens when the server updates while a download is going on because of this. It can cause certain files to be out-of-date middownload. This could be a silent killer, causing users to see that randomly files are out of date, or that files randomly fail signature checks, meaning they need to start all the way at the beginning. The solution to this would be to store atomic files (I'll call them objects) at unique directories. In the manifest there would be the path and the remote file. Say if path = gameinfo.txt,  object = uuid,  it would download the file at /launcher/objects/uuid and save it to /gameinfo.txt. This also has the benefit of allowing downgrades.
> * An MD5 hash could be used to verify the integrity of the files on disk, and would download the file (with the SHA384 being verified at download time).

If this is issue-spam I want as much of it as possible. I'm relatively new to writing software for a real purpose so any info on something that I haven't realised is greatly appreciated.

Anyway - 
1) I did think about json but one of the other devs kinda shot it down - I'll take another look into it.
2) Signing all the files does reduce the chance of a single POF incase the packer fails to sign it but I don't think it's worth signing if that's the only benefit.
3) This seems a lot better - but I don't really get the MD5 bit you're talking about - if you could elaborate that'd be appreciated.

Also I'm planning on modularising atomic as at the moment the currrent plan of just throwing everything in isn't really going to be efficient this module will be called `ofdl` with other modules such as `ofverify` and anything else needed with atomic being a collection of those modules most likely called `oftools` or something to that effect.

**Comment by w3bb** [comment]
@int-72h MD5 is fast but bad for checking if a file had been intentionally tampered with, and SHA-384 is is slow but good for checking if the file had been tampered with (which is why its used in signature checks). If file verification were to be done (as in verifying the integrity of existing, downloaded files), we shouldn't be using SHA-384 to check for incidential or non-malicious corruption.

If it were to be done, an ideal flow would be something like this. local = file on disk, db = manifest info on file(pseudocode)

```
if db.md5 hash doesn't match local.md5 {
 downloadAndVerifySHA384(local.file)
} else {
 return
}
```

The reason we only verify it after download is that the place where intentional tampering could have occured is when it's downloaded from the internet. However if it's on the disk we should assume it hasn't been tampered with.

Of course this is a sort of feature proposal, but having verification/integrity check functionality would be useful outside of just manual verification (also because I want to implement this in murse).

**Comment by int-72h** [comment]
Ah, I get you- that's going to be in ofverify which I'll get to writing once this is less ..buggy

**Comment by w3bb** [comment]
I wrote a little program that converts the ofatomic sqlite database to a JSON format (in these tests I used a condensed/ugly version of the JSON, which is what you would send over the wire), and used a tool that converts SQLite databases to csv. Here's my results:

File sizes:
JSON: 5.3M
CSV: 4.7M
SQLite: 6.3M

File sizes after ztd compression:
JSON: 2.1M
CSV: 2.1M
SQLite: 2.6M

For the CSV I used some basic online tool. It wasn't a particularly pretty or useful format. It's more of a ballpark estimate of the size. So pretty much it just boils down to whether CSV or JSON decode better. Unfortunately I can't benchmark CSV because the schema for the manifest would have to be completely redone, but I was able to get the decoding performance of  JSON using encoding/json from Go's standard library. 

JSON:  47ms

Oh yeah, I guess that's a good point to bring up. CSV is going to be a pain in the ass to read and doesn't have much of a benefit over JSON. You can't represent data the same way you can in JSON/SQLite, so you're going to have to change up the format.

[JSON (Pretty)](https://files.catbox.moe/40i3om.json)
[JSON (Ugly)](https://files.catbox.moe/myvlgj)
[CSV (missing metadata)](https://files.catbox.moe/wtu6tz.csv)

**Comment by int-72h** [comment]
Very convicing points, I'll bring it up with the team.

**Comment by int-72h** [comment]
OK the general gist I'm getting is "leave the current protocol" since we're getting major feature creep which stopped the launchers from coming out any sooner. Since I do believe to some extent time is important here but I really want to implement these ideas I'm probably going to split atomic into just a set of command line tools and then a testing "suite" for the new protocol. The atomic name will probably be retired as the new command line tools will be called ofdtools and the protocol tester ofptests or something to that effect. This'll also give me more freedom to entirely redesign the system hopefully with less "bloat", taking your ideas of uuids and atomic operations (...with *atomic* operations probably being a major reason I think it's a good idea I'm renaming this)

**Comment by w3bb** [comment]
> we're getting major feature creep

Can you elaborate on this? In terms of the protocol it's extremely simple.

**Comment by w3bb** [comment]
https://paste.sr.ht/~welt/33dbabe5bf7c541e6b29b36d203d028b809c960d Here's a first draft for the alternative API I'm working on, if you're looking for inspiration.

**Comment by int-72h** [comment]
> > we're getting major feature creep
> 
> Can you elaborate on this? In terms of the protocol it's extremely simple (aside from SQLite)

The launcher the team want to get out is Stainless, which is one written in c++, meant for general users and currently only has signing support missing - When I joined launcher development I pushed for more changes to stainless like zstd and signing and used atomic just as a testbed for these since I'm not amazingly proficient in C++/ any other language (I do like how murse is written in golang, good choice) - After I made atomic into something for all command line users I wanted to optimise the protocol further, but the devs wanted to prioritise stainless's release since theres still quite a few QOL changes that need to be made on top of signing, and at this time that does seem more pressing than reimplementing the protocol before release. Since I'm bogged with college at the moment atomic's split and v2 development will be probably a month out.

*tl;dr They really just want to finish stainless and release it ASAP.*

**Comment by w3bb** [comment]
> > > we're getting major feature creep
> > 
> > 
> > Can you elaborate on this? In terms of the protocol it's extremely simple (aside from SQLite)
> 
> The launcher the team want to get out is Stainless, which is one written in c++, meant for general users and currently only has signing support missing - When I joined launcher development I pushed for more changes to stainless like zstd and signing and used atomic just as a testbed for these since I'm not amazingly proficient in C++/ any other language (I do like how murse is written in golang, good choice) - After I made atomic into something for all command line users I wanted to optimise the protocol further, but the devs wanted to prioritise stainless's release since theres still quite a few QOL changes that need to be made on top of signing, and at this time that does seem more pressing than reimplementing the protocol before release. Since I'm bogged with college at the moment atomic's split and v2 development will be probably a month out.
> 
> _tl;dr They really just want to finish stainless and release it ASAP._

Ah okay, interesting bit of information. Thanks for sharing.

**Comment by int-72h** [comment]
> > > > we're getting major feature creep
> > > 
> > > 
> > > Can you elaborate on this? In terms of the protocol it's extremely simple (aside from SQLite)
> > 
> > 
> > The launcher the team want to get out is Stainless, which is one written in c++, meant for general users and currently only has signing support missing - When I joined launcher development I pushed for more changes to stainless like zstd and signing and used atomic just as a testbed for these since I'm not amazingly proficient in C++/ any other language (I do like how murse is written in golang, good choice) - After I made atomic into something for all command line users I wanted to optimise the protocol further, but the devs wanted to prioritise stainless's release since theres still quite a few QOL changes that need to be made on top of signing, and at this time that does seem more pressing than reimplementing the protocol before release. Since I'm bogged with college at the moment atomic's split and v2 development will be probably a month out.
> > _tl;dr They really just want to finish stainless and release it ASAP._
> 
> Ah okay, interesting bit of information. Thanks for sharing.

I appreciate your feedback with the protocol and indirectly showing me to sourcehuts existence - May start using this for smaller projects of mine, very nice - and I'll definitely start drafting the next protocol within the coming weeks when I get some time since a low overhead protocol (theoretically low overhead anyway) is my college project, actually.

**Comment by w3bb** [comment]
> > > > > we're getting major feature creep
> > > > 
> > > > 
> > > > Can you elaborate on this? In terms of the protocol it's extremely simple (aside from SQLite)
> > > 
> > > 
> > > The launcher the team want to get out is Stainless, which is one written in c++, meant for general users and currently only has signing support missing - When I joined launcher development I pushed for more changes to stainless like zstd and signing and used atomic just as a testbed for these since I'm not amazingly proficient in C++/ any other language (I do like how murse is written in golang, good choice) - After I made atomic into something for all command line users I wanted to optimise the protocol further, but the devs wanted to prioritise stainless's release since theres still quite a few QOL changes that need to be made on top of signing, and at this time that does seem more pressing than reimplementing the protocol before release. Since I'm bogged with college at the moment atomic's split and v2 development will be probably a month out.
> > > _tl;dr They really just want to finish stainless and release it ASAP._
> > 
> > 
> > Ah okay, interesting bit of information. Thanks for sharing.
> 
> I appreciate your feedback with the protocol and indirectly showing me to sourcehuts existence - May start using this for smaller projects of mine, very nice - and I'll definitely start drafting the next protocol within the coming weeks when I get some time since a low overhead protocol (theoretically low overhead anyway) is my college project, actually (granted entirely unrelated and most likely inefficient).

Just keep in mind once Sourcehut leaves alpha it will be paid only. And recently the build system (which rocks by the way) is paid only due to abuse. Although two bucks a month is basically nothing.

**Comment by int-72h** [comment]
> > > > > > we're getting major feature creep
> > > > > 
> > > > > 
> > > > > Can you elaborate on this? In terms of the protocol it's extremely simple (aside from SQLite)
> > > > 
> > > > 
> > > > The launcher the team want to get out is Stainless, which is one written in c++, meant for general users and currently only has signing support missing - When I joined launcher development I pushed for more changes to stainless like zstd and signing and used atomic just as a testbed for these since I'm not amazingly proficient in C++/ any other language (I do like how murse is written in golang, good choice) - After I made atomic into something for all command line users I wanted to optimise the protocol further, but the devs wanted to prioritise stainless's release since theres still quite a few QOL changes that need to be made on top of signing, and at this time that does seem more pressing than reimplementing the protocol before release. Since I'm bogged with college at the moment atomic's split and v2 development will be probably a month out.
> > > > _tl;dr They really just want to finish stainless and release it ASAP._
> > > 
> > > 
> > > Ah okay, interesting bit of information. Thanks for sharing.
> > 
> > 
> > I appreciate your feedback with the protocol and indirectly showing me to sourcehuts existence - May start using this for smaller projects of mine, very nice - and I'll definitely start drafting the next protocol within the coming weeks when I get some time since a low overhead protocol (theoretically low overhead anyway) is my college project, actually (granted entirely unrelated and most likely inefficient).
> 
> Just keep in mind once Sourcehut leaves alpha it will be paid only. And recently the build system (which rocks by the way) is paid only due to abuse. Although two bucks a month is basically nothing.

I see - didn't know that but 24 a year isn't much at all (considering the prices of most subscriptions...)
Anyway I'll hopefully get no compression as an option before release.

**Comment by w3bb** [comment]
> > > > > > > we're getting major feature creep
> > > > > > 
> > > > > > 
> > > > > > Can you elaborate on this? In terms of the protocol it's extremely simple (aside from SQLite)
> > > > > 
> > > > > 
> > > > > The launcher the team want to get out is Stainless, which is one written in c++, meant for general users and currently only has signing support missing - When I joined launcher development I pushed for more changes to stainless like zstd and signing and used atomic just as a testbed for these since I'm not amazingly proficient in C++/ any other language (I do like how murse is written in golang, good choice) - After I made atomic into something for all command line users I wanted to optimise the protocol further, but the devs wanted to prioritise stainless's release since theres still quite a few QOL changes that need to be made on top of signing, and at this time that does seem more pressing than reimplementing the protocol before release. Since I'm bogged with college at the moment atomic's split and v2 development will be probably a month out.
> > > > > _tl;dr They really just want to finish stainless and release it ASAP._
> > > > 
> > > > 
> > > > Ah okay, interesting bit of information. Thanks for sharing.
> > > 
> > > 
> > > I appreciate your feedback with the protocol and indirectly showing me to sourcehuts existence - May start using this for smaller projects of mine, very nice - and I'll definitely start drafting the next protocol within the coming weeks when I get some time since a low overhead protocol (theoretically low overhead anyway) is my college project, actually (granted entirely unrelated and most likely inefficient).
> > 
> > 
> > Just keep in mind once Sourcehut leaves alpha it will be paid only. And recently the build system (which rocks by the way) is paid only due to abuse. Although two bucks a month is basically nothing.
> 
> I see - didn't know that but 24 a year isn't much at all (considering the prices of most subscriptions...)
> Anyway I'll hopefully get no compression as an option before release.
Not even, it's $20. I don't know how Drew is able to afford anything for Sourcehut.

**Comment by int-72h** [comment]
...Ok thats enough quoting. Only thing that I'd miss would be issues but I think a simple mirror to GH will do the trick.

**Comment by w3bb** [comment]
> ...Ok thats enough quoting. Only thing that I'd miss would be issues but I think a simple mirror to GH will do the trick.

I might have a bad habit from email...

https://github.com/aThorp96/sourcehut_issue_mirror

Todo is the equivalent of issues. You can see my todo page [here](https://todo.sr.ht/~welt/murse) for example.

**Comment by int-72h** [comment]
oh * thats neat*. Quite like the design of sourcehut as well.

**Comment by int-72h** [comment]
Confimed with the devs that uncompressed will be available on the backend - closing untill they implement it on their side first.
